### DISCUSSION TITLE
How to map missing lower and greater than symbols on an azerty thinkpad T430 keyboard?

### DISCUSSION THREAD

**Discussion by jeanpat** [houmain/keymapper#15] [Q&A]
Hello,
I have a thinkpad T430, azerty keyboard, running endeavourOS. I have been searching the way to type the lower and greater than symbol and I didn't find. 
So I'm trying to remap these two symbols using alt+w and alt+x, as follow, first I wrote a keymapper.conf file containing:
```
# define some more aliases
  Alt   = AltLeft
  AltGr = AltRight
  Win   = Meta

# Mapping on thinkpad T430 azerty keyboard

Alt{w} >> <
Alt{x} >> >
```
Then from a terminal:
```bash
$ systemctl start keymapperd
[jeanpat@T430 ~]$ keymapper
Invalid key 'w' in line 8
Loading configuration failed
```
Could you tell me how to modify the file to get the mapping?
Thanks.

**Comment by houmain** [comment]
Hi! 
This keyboard https://www.s2i33.com/claviers-portable/5879-clavier-azerty-lenovo-t430-fru04x1288-gar3-mois.html ? 
The </> key is called IntlBackslash. Is it missing on your keyboard?
The configuration file is case-sensitive, that's why it complains about the 'w'. This configuration should work:
```
# define some more aliases
  Alt   = AltLeft
  AltGr = AltRight
  Win   = Meta

# Mapping on thinkpad T430 azerty keyboard

Alt{W} >> !Alt IntlBackslash
Alt{X} >> !Alt Shift{IntlBackslash}
```
On my Linux machine it unfortunately does not! Under Windows it works as expected. 
There seems to be a problem with the Alt key I need to investigate. So thanks for pointing that out!

**Reply by houmain** [reply]
The problem is that Alt is always immediately applied (like Shift and Ctrl, since they are also used as mouse modifiers) and in most applications this sets the focus on the menu bar, where a following < or > does nothing. I will try to come up with a solution to this problem.

**Comment by jeanpat** [comment]
This is not exactly the same keyboard: https://ibb.co/3c367Xs

**Reply by houmain** [reply]
Hi Jean, 
with commit 5f4aec449e0435bd32d4cd116d363979658b0bcc Alt is no longer immediately applied and this configuration should work:
```
# define some more aliases
Alt   = AltLeft
AltGr = AltRight
Win   = Meta

# Mapping on thinkpad T430 azerty keyboard
Alt{W} >> IntlBackslash
Alt{X} >> Shift{IntlBackslash}
```
Did you try a binary release or are you building from source? I have not created a new release yet.

**Reply by jeanpat** [reply]
Hello,
Sorry for delayed answer.
I have updated kemapper as follow:
`yay -S keymapper`
The configuration file was just copied from your last message.
keymapperd was started:
`$ systemctl start keymapperd`
then keymapper was started in verbose mode:
```
$ keymapper -v
Loading configuration file '/home/jeanpat/.config/keymapper.conf'
Connecting to keymapperd
Initializing focused window detection
Entering update loop
```
At this point all the keyboard  (in gnome) stop to work. Closing the terminal re-enables it
[2021-12-04 11-44-38.zip](https://github.com/houmain/keymapper/files/7653796/2021-12-04.11-44-38.zip)
.
I guess, I should have reinstall from source.

**Reply by houmain** [reply]
Hi, 
there is also a `keymapper-git` package in the AUR. The `keymapper` project was not updated yet. 
But installing `keymapper-git` instead of `keymapper` should just fix the mentioned problem with mapping the `Alt` key.
With this configuration your keyboard should not stop working with any version! (One could block everything by adding the mapping `Any >> `).
Do you have any other input devices connected or installed any special software which could interfere? I know that `usbguard` can be used to block unknown USB devices...
Could you please also start keymapperd with the -v parameter (simply with `sudo keymapperd -v`). It should output the devices it grabbed...
Thanks in advance!

**Reply by jeanpat** [reply]
Hi,
keymapper 1.6 was uninstalled:
`$ yay -R keymapper`
and the git version is installed
```
$ yay keymapper-git
1 aur/keymapper-git latest-2 (+1 0.16) (Installé : r97.7c092b1-1)
    A cross-platform context-aware key remapper.
==> Paquets à installer (ex. 1 2 3, 1-3 or ^4)
```

Now starting the daemon as follow inactivates the keyboard (need to restart):
`$ systemctl start keymapperd`
It doesn't seem possible to pass a verbose mode:
```
$ systemctl start keymapperd -v
systemctl : option invalide -- 'v'
```

**Reply by houmain** [reply]
Sorry, that it is causing so much trouble on your machine.
I know what happened. The package installer adds `keymapper` to your autostarted applications. I does nothing as long as keymapperd is not running. But as soon as you started the `keymapperd` service your keyboard locked up :( .

You should stop the autostarted keymapper first:
`killall keymapper`
Then you can start keymapperd directly without the `systemctl start`, just using sudo. So in one terminal:
`sudo keymapperd -v`
and in another:
`keymapper`
Closing the second terminal should re-enable your keyboard and permit to paste a copy of keymapperd's output here.
I am not sure, that it will output something valuable, but at the moment I am still clueless what is happening. Thanks!

**Reply by jeanpat** [reply]
Not that much trouble, thank you for your help and time.

**Comment by KaffeMyers** [comment]
I seem to have the same issue.
```
Waiting for keymapper to connect
Creating uinput keyboard 'Keymapper'
Updating device list
Grabbing device event0 'AT Translated Set 2 keyboard'
Entering update loop
```

This is all that happens and my keyboard dies entirely.

The config I'm trying with is a simple `A >> B`

I'm running this on Debian Bookworm, on a HP EliteBook 840 G3. I use X, but no fancy desktop environment. I tried shutting down sxhkd to see if it interfered somehow, but no change. I installed the latest prepackaged build (keymapper 1.6.0)

**Reply by houmain** [reply]
Hi, thank you very much for your thorough report. 
Fortunately I could reproduce this behavior now in a virtual machine! (with the EndeavourOS ISO).

**Reply by houmain** [reply]
The last commit (f09c49cabbdf9060b72ef5b3f92941a6d83f4063) fixed the problem in my test setup. Could you please verify that it resolves the issue also on your machine? (building from source)

**Reply by KaffeMyers** [reply]
Seems to be working for me! Thanks for the awsome work and the swift resolution.

**Reply by houmain** [reply]
Great! 
The created virtual keyboard previously advertised that it has all possible keys (over 700). I could imagine that some commonly installed application exclusively grabs devices with some very exotic key.
